# Pregunta 9.Estamos pensando en abrir una nueva vertical llamada Mercadona eat&go. Su objetivo es que la gente pueda comprar y comer los productos en una sala acondicionada dentro de nuestro establecimiento. 
# Te compartimos un csv con las compras de algunos usuarios en verano. ¿Qué tipo de alimentos recomendarías y por qué?


### Punto a tomar en cuenta:
* Las compras fueron realizadas en verano, esto quiere decir que el tipo de comida para verano suele ser mas fresca y liviana.

* Se tiene que tomar en cuenta tambien que sean alimentos que puedan consumirse en cualquier estación del año.

###  Qué se hará?
* 📌 En primer lugar, se generará un data frame con las compras y los nombres de 

In [22]:
import pandas as pd
import numpy as np
import io
pd.set_option('display.float_format', lambda x: '%.2f' % x)
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
rd = pd.read_csv('raw_data.csv')

In [3]:
dfc = pd.read_csv('df_compras_100000.csv')

In [4]:
rd.head(5)

category  \
0  perfumeria_e_higiene_cuidado_facial_protector_...   
1      perfumeria_e_higiene_higiene_intima_compresas   
2      perfumeria_e_higiene_higiene_intima_compresas   
3  perfumeria_e_higiene_higiene_bucal_productos_p...   
4  perfumeria_e_higiene_higiene_bucal_productos_p...   

                                                name description  price  \
0     LIPOSAN protector labial soft rose envase 1 ud         NaN   2.39   
1         AUSONIA compresas normal alas bolsa 14 uds         NaN   2.49   
2  EVAX Cottonlike compresa super plus con alas b...         NaN   2.79   
3  POLIDENT tabletas limpiadoras para prótesis de...         NaN   7.95   
4  IWHITE kit de blanquamiento dental profesional...         NaN  36.99   

   reference_price reference_unit          insert_date  \
0             2.39             ud  2021-03-06 00:00:00   
1             0.18             ud  2021-03-06 00:00:00   
2             0.23             ud  2021-03-06 00:00:00   
3             0.11             ud  2021-03-06 00:00:00   
4             3.70             ud  2021-03-06 00:00:00   

                         product_id id_reference_supermarket  
0  45d83ce427cf60c02e1e6f3d28f58ed6                       M3  
1  3fa10738dcbe5f09646eaf39a46a2871                       M3  
2  5f5bc0a466375a16d01a6aed6c66e413                       M3  
3  0a1e56be01f8766d0bb87f3e42fd2abf                       M3  
4  b150c5e27da5c91d13a5b570c9b3efd3                       M3

In [5]:
dfc

id_compra fecha_compra  id_cliente                       id_producto
0              0   2022-06-02          19  f65b04468e58cc9f604cb097236872fe
1              1   2022-08-02          27  ce43ded19a577ac0aaa1e3b0cf2b5f1a
2              2   2022-07-09          20  0a56c32642021feae8049bbb840a6153
3              3   2022-08-18           5  2f8af1c55f93e0b3e4a0b03e5e0f28b4
4              4   2022-06-27           3  d7382d624ca08e4459996938e65935e8
...          ...          ...         ...                               ...
99995      99995   2022-07-19          27  e98212fd3744bc61ee8664b78381f592
99996      99996   2022-08-04          50  30c0d240eb3834a32a4ca8128fa2a546
99997      99997   2022-07-11          20  b408c71cf823b99ff6b80e17573d5940
99998      99998   2022-07-09           1  beadab32512739ed268a6c4dcf19947e
99999      99999   2022-06-27          29  0df17d9bc9b3a7846a4527d8ec94090a

[100000 rows x 4 columns]

In [24]:
len(dfc)

100000

# En el dataframe de compras (dfc) vemos que un cliente puede realizar varias compras en diferentes días. Se realizará lo siguiente:

* Valoramos por medio de un resumen de la base de datos dfc para conocer cual es el id que mas se repite.
* Filtraremos la base de datos rd por M1 (Mercadona)
* Como sólo necesitamos dos columnas del df de Mecadona, aplicamos un filtro para traernos las columnas **product_id y name**

In [7]:
dfc.id_producto.describe()   #vemos que el id_producto que mas se repite es el "93ca9085566a3803e9538c1b756204b5"

count                               100000
unique                                 200
top       93ca9085566a3803e9538c1b756204b5
freq                                   543
Name: id_producto, dtype: object

## Creamos el df con las filas que necesitamos del rd

In [85]:
rd_filas = rd.groupby(["product_id", "name","id_reference_supermarket"]).agg( conteo_en_df = ("product_id" , "count")).reset_index() #cogemos solo las columnas que necesitamos de la df de Mercadona

In [86]:
rd_filas

product_id  \
0      000524d73b9d09ae193aaf463191b038   
1      000541efc0b1576743bc990d241aaa52   
2      0007507cdc06cdf416c777d38d69eb14   
3      000780b12d67871681dccc1383869831   
4      0008afbaae96e6003d71de151a2ec67b   
...                                 ...   
40432  fff3dd4faa8b2be5178fab66d333acd9   
40433  fff473551393a45da35c7ea160c52ada   
40434  fff6e6b26b54a9fb3cf990f0a967eadd   
40435  fff8007d5d155de4a4ea5215b0a0ffc5   
40436  fff8007d5d155de4a4ea5215b0a0ffc5   

                                                    name  \
0      Pechuga de pavo Braseado Campofrío Finissimas ...   
1      Cerveza El Águila sin filtrar pack de 10 latas...   
2                            Mopa de Microfibra ROZENBAL   
3              Caramelos de goma Favoritos Haribo 275 g.   
4      DIA MARI MARINERA medallones de merluza del ca...   
...                                                  ...   
40432  Yogur bífidus desnatado con kiwi Danone Activi...   
40433         Bebida láctea de fresa y plátano Hacendado   
40434  Tortitas de arroz  integrales con chocolate y ...   
40435                          Mix de verduras jardinera   
40436               Mix de verduras jardinera para micro   

      id_reference_supermarket  conteo_en_df  
0                           M2           167  
1                           M2             3  
2                           M2             9  
3                           M2           139  
4                           M3           255  
...                        ...           ...  
40432                       M2            21  
40433                       M1           559  
40434                       M2            12  
40435                       M1           541  
40436                       M1            18  

[40437 rows x 4 columns]

## Filtramos por Supermercado, en este caso Mercadona

In [84]:
rd_solo_mercadona= rd_filas[(rd_filas["id_reference_supermarket"] == "M1")]# filtramos por supermercado
rd_solo_mercadona

product_id  \
9      000d8079b6a0daae2b0267126ae9adbf   
10     0010432366ece6221db98d6f1dc15d0b   
15     0015c4bf6eb2d44a1ebb32f0999a2e2b   
24     0023b050194a9f4b764860f3e563a9c5   
27     002c7f51dd28e4f640ae52f12dfe1f1a   
...                                 ...   
40424  ffe0e8d74c9dfdae53981b4d0dfd72ed   
40427  ffed227a44b6eb41d6badee80731470a   
40433  fff473551393a45da35c7ea160c52ada   
40435  fff8007d5d155de4a4ea5215b0a0ffc5   
40436  fff8007d5d155de4a4ea5215b0a0ffc5   

                                                    name  \
9               Comprimidos Vitamina D3 2000 ui Vitadiet   
10     Vasos grandes de plástico Bosque Verde reutili...   
15     Baño de color 210 negro azulado Deliplus 0% am...   
24     Mini eau de parfum hombre Flor de Mayo Triomph...   
27                         Queso rallado Gouda Hacendado   
...                                                  ...   
40424               Eau de toilette mujer Deliplus Vuela   
40427            Helado cucurucho de chocolate Hacendado   
40433         Bebida láctea de fresa y plátano Hacendado   
40435                          Mix de verduras jardinera   
40436               Mix de verduras jardinera para micro   

      id_reference_supermarket  conteo_en_df  
9                           M1           243  
10                          M1            88  
15                          M1           438  
24                          M1           139  
27                          M1           157  
...                        ...           ...  
40424                       M1           216  
40427                       M1            92  
40433                       M1           559  
40435                       M1           541  
40436                       M1            18  

[9904 rows x 4 columns]

# Se tiene una base de datos de las ventas de Junio, Julio y Agosto, con los id de los productos, pero no sabemos los nombres de estos. Que haremos?

Necesitamos unir los dos dataframe por medio de los id de los productos, de esta manera podemos conocer el nombre del producto que se ha vendido en ese trimestre.

In [92]:
Merge_dfs = dfc.merge(rd_solo_mercadona,how="left" , left_on ="id_producto" , right_on = "product_id" )
Merge_dfs

id_compra fecha_compra  id_cliente                       id_producto  \
0               0   2022-06-02          19  f65b04468e58cc9f604cb097236872fe   
1               1   2022-08-02          27  ce43ded19a577ac0aaa1e3b0cf2b5f1a   
2               2   2022-07-09          20  0a56c32642021feae8049bbb840a6153   
3               3   2022-08-18           5  2f8af1c55f93e0b3e4a0b03e5e0f28b4   
4               4   2022-06-27           3  d7382d624ca08e4459996938e65935e8   
...           ...          ...         ...                               ...   
102490      99995   2022-07-19          27  e98212fd3744bc61ee8664b78381f592   
102491      99996   2022-08-04          50  30c0d240eb3834a32a4ca8128fa2a546   
102492      99997   2022-07-11          20  b408c71cf823b99ff6b80e17573d5940   
102493      99998   2022-07-09           1  beadab32512739ed268a6c4dcf19947e   
102494      99999   2022-06-27          29  0df17d9bc9b3a7846a4527d8ec94090a   

                              product_id                     name  \
0                                    NaN                      NaN   
1                                    NaN                      NaN   
2                                    NaN                      NaN   
3                                    NaN                      NaN   
4                                    NaN                      NaN   
...                                  ...                      ...   
102490                               NaN                      NaN   
102491                               NaN                      NaN   
102492                               NaN                      NaN   
102493  beadab32512739ed268a6c4dcf19947e  Refresco cola Hacendado   
102494                               NaN                      NaN   

       id_reference_supermarket  conteo_en_df  
0                           NaN           NaN  
1                           NaN           NaN  
2                           NaN           NaN  
3                           NaN           NaN  
4                           NaN           NaN  
...                         ...           ...  
102490                      NaN           NaN  
102491                      NaN           NaN  
102492                      NaN           NaN  
102493                       M1        559.00  
102494                      NaN           NaN  

[102495 rows x 8 columns]

In [93]:
Merge_dfs.name.describe()

count                  32129
unique                    24
top       Refresco Coca-Cola
freq                    3492
Name: name, dtype: object

# Vamos a agrupar los df unidos para observar cual es el producto que mas se vende

In [122]:
nombre_producto = Merge_dfs.groupby(["id_producto","name"]).agg(producto_comprado = ("name", "count")).reset_index()


In [123]:
nombre_producto

id_producto  \
0   046d8a822e8835a10fd34d3a50fa4ac3   
1   06c2cd4a8cc5383dcf9dfb7d0612ade1   
2   0d6ecab1dea5b0caf81cb81fcf66a148   
3   0fee679e6cd7841821c432b1fa529ccc   
4   18744a79b1f1615f30b3d54554dd4500   
..                               ...   
59  ee59c29d26690c3dcc30c9683646d301   
60  f53c78cdffb3053b1948ffdb067f48e7   
61  f56e65022625f37a24a3630410c5c071   
62  f5e9f981a029e446490348d73789c847   
63  f6a62a8c304a0e818f869aa2ea02df8a   

                                               name  producto_comprado  
0                           Refresco cola Hacendado                512  
1                    Refresco Coca-Cola Zero azúcar                494  
2                   Bebida isotónica blue Iso drink                519  
3                           Refresco cola Hacendado                490  
4                    Refresco Coca-Cola Zero azúcar                513  
..                                              ...                ...  
59                     Refresco Coca-Cola Zero Zero                539  
60                   Refresco Coca-Cola Zero azúcar                522  
61                  Bebida energética original Burn                490  
62                     Refresco Coca-Cola Zero Zero                518  
63  Bebida para deportistas sabor cítrico Hacendado                515  

[64 rows x 3 columns]

# Ahora bien, por alguna razón solo aparecen bebidas. ¿Que pasa con los alimentos?

Estudiaremos los id_productos de la df original (rd) para ver si pertenecen a los de mercadona,  porque puede que esté pasando que algunos ids pertenezcan a otro supermercado.

In [119]:
prueba1= rd[(rd["product_id"]== "f65b04468e58cc9f604cb097236872fe")] 
# EL product_id qeu aparece en dfc no es de mercadona sino de dia 
prueba.head(2)

category  \
770    alimentacion_caldos_sopas_y_pures_caldo_concen...   
29263  alimentacion_caldos_sopas_y_pures_caldo_concen...   

                                                    name description  price  \
770    KNORR doble caldo de carne estuche 24 pastilla...         NaN   2.89   
29263  KNORR doble caldo de carne estuche 24 pastilla...         NaN   2.89   

       reference_price reference_unit          insert_date  \
770               0.12             ud  2021-03-06 00:00:00   
29263             0.12             ud  2021-03-07 00:00:00   

                             product_id id_reference_supermarket  
770    f65b04468e58cc9f604cb097236872fe                       M3  
29263  f65b04468e58cc9f604cb097236872fe                       M3

In [120]:
prueba2= rd[(rd["product_id"]== "f56e65022625f37a24a3630410c5c071")] 
# EL product_id qeu aparece en dfc no es de mercadona sino de dia 
prueba.head(2)

category  \
770    alimentacion_caldos_sopas_y_pures_caldo_concen...   
29263  alimentacion_caldos_sopas_y_pures_caldo_concen...   

                                                    name description  price  \
770    KNORR doble caldo de carne estuche 24 pastilla...         NaN   2.89   
29263  KNORR doble caldo de carne estuche 24 pastilla...         NaN   2.89   

       reference_price reference_unit          insert_date  \
770               0.12             ud  2021-03-06 00:00:00   
29263             0.12             ud  2021-03-07 00:00:00   

                             product_id id_reference_supermarket  
770    f65b04468e58cc9f604cb097236872fe                       M3  
29263  f65b04468e58cc9f604cb097236872fe                       M3

# Queda demostrado que el id del producto de la base de datos dfc (compras) pertenece a otro supermercado, en este caso a DIA


In [192]:
df_np = Merge_dfs.groupby(["id_producto","name"]).agg(compras_realizadas =("id_producto","count")).sort_values( by = ["compras_realizadas"] , ascending=False)

In [193]:
df_np

compras_realizadas
id_producto                      name                                                            
ee59c29d26690c3dcc30c9683646d301 Refresco Coca-Cola Zero Zero                                 539
2296096c9fa43436a6e987412e7eda0a Bebida energética Red Bull                                   537
4d14041baa4d7611411b353fcc6975cb Refresco cola Pepsi                                          533
a5101119b3a8b1c266db97f44113fd09 Bebida energética sabor guaraná Energy Drink                 530
451a3c6251a23a555066da9c472ca423 Refresco Coca-Cola Zero Zero                                 528
...                                                                                           ...
289719fb2a07d2cbd18fb5724fffa747 Bebida isotónica de naranja Iso drink                        469
7a497a5a45597bee31bf2d05acecd999 Bebida energética energy ultra zero Monster                  466
                                 Bebida energética energy ultra Monster                       466
                                 Bebida energética Energy Ultra zero Monster                  466
b2d8c59e8ca2bbd36e41c55da3ee4813 Bebida energética Energy Drink                               458

[64 rows x 1 columns]

In [156]:
Merge_dfs.name.describe()

count                  32129
unique                    24
top       Refresco Coca-Cola
freq                    3492
Name: name, dtype: object

# Respuesta PREGUNTA 9

## ARGUMENTO

En principio contabamos con DataFrame de compras realizadas en el mes de Junio, Julio y Agosto en el Mercadona, esta estaba compuesta por las columnas **id_compra,fecha_compra, id_cliente, id_producto** y **100.000 registros** pero no contabamos con lo nombres de los productos, lo cual necesitabamos para conocer los articulos mas vendidos y poder sacar posibles platos.

* En primer lugar se trajo las columnas de la base de datos original *(raw_data)* para poder hacer la union de las tablas por medio de los id del producto.
* Se realizó la unión de la tabla de compras, con la base de datos original, filtrandolo por el supermercado Mercadona creando un nuevo df con **102495 registros**
* Se notó de inmediato que no habian registros con los nombres de los alimentos.
* Se realizó unas pruebas para ver cual era la razón por la cual no aparecian estos nombres en la union de la base de datos.


# CONCLUSIÓN

Se observó que de las 102.495 registros coincidian solo **64 registros** y todos estos haciendo referencia a bebidas,al comparar los id de los productos con la base de datos original, estos coincidian con el id de otros supermercados. Por ejemplo el id **"f65b04468e58cc9f604cb097236872fe"** que se encontraba en la base de datos de las compras, **pertenecía a DIA**. Se aplicó un filtro con los id que solo pertenecieran a Mercadona. 

No se puede recomendar ningun alimento excepto bebidas ya que no contamos con registros de compras de alimentos para sacar una conclusion, como bebida podemos recomendar la **Refresco Coca-Cola Zero Zero =539 compras,Bebida energética Red Bull=537 compras y Refresco cola Pepsi=533 compras** sabemos con certeza que al menos el Refresco Coca-Cola Zero Zero no puede faltar ademas que es bebible todas las estaciones del año.	